qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -365,10 +419,46 @@ def _circuit_u(theta,
                    phase,
                    simplify=True,
                    atol=DEFAULT_ATOL):
-        # pylint: disable=unused-argument
         qr = QuantumRegister(1, 'qr')
         circuit = QuantumCircuit(qr, global_phase=phase)
-        circuit._append(UGate(theta, phi, lam), [qr[0]], [])
+        if not simplify:
+            atol = -1.0
+        if abs(theta) > atol or abs(phi) > atol or abs(lam) > atol:
+            circuit._append(UGate(theta, _mod_2pi(phi, atol), _mod_2pi(lam, atol)), [qr[0]], [])
+        return circuit
+
+    @staticmethod
+    def _circuit_psx_gen(theta, phi, lam, phase, atol, pfun, xfun, xpifun=None):
```

Comment:
I agree that what you propose is worth considering as a final landing place (ie I mean I'm not sure yet which is better, and I need to think about it more although currently I lean towards the current approach). This PR already grew in scope like 3 orders of magnitude past what I originally planned, and I don't want to make it even larger by introducing new passes, corresponding passmanager defaults (and associated tests), etc so I stuck with making the existing functionality correct, within tolerances, and properly covered by test cases (and then folding in the zsxx because otherwise that would have to be done twice). Previously there were situations where one out of PSX, ZSX was correct and the other incorrect, etc, at least they are all consistent once this merges.

The cases which are most obviously candidates for combining are:
 - PSZ, ZSX, U1X, ZSXX (already combined under _psx_gen)
 - U, U3, U321 (but they are already very simple so not much benefit to combining)
 - ZYZ, ZXZ, XYX (could be an _euler_gen with parameters of the 2 orthogonal gates as with _psx_gen, or more like your proposal implemented as UZYZUdag where U is a basis rotation into ZYZ basis, and then a follow up pass changes UZYZU into eg XYX)

One downside with pushing final basis selection to a later pass is that it means another walk over the dag, and although mostly just local fixes (eg U1X->PSX has replace all U1 with P and update global phase) at least in the case of ZSXX means searching for a pattern of dagnodes SX-RZ(alpha)-SX where`abs(_mod_2pi(alpha))<tol`, so this would be an expensive way to do this versus just generating the simplified pattern in the first place and will be worth profiling.

Can we defer possible such refactoring as a follow-up patch after merging this one?